### PR TITLE
Improve handling of GC managed array buffers

### PR DESCRIPTION
**Pull request by tschneidereit**
Specifically, some of the `noGC` guards had issues that could lead to panics during exception throwing, and some of the loops over multiple buffers where overly complex.

### CODE REVIEW

**Comment by tschneidereit** [comment]
I'll change this to target `main` before landing, and will wait for the `fetch-rework` branch to land before doing so.

**Review by guybedford — APPROVED** [review]

**Pull request merged**
